test/wdio/slot-html/cmp.test.tsx

```diff
@@ -0,0 +1,186 @@
+    await expect(results6DefaultSlotChildren[1]).toHaveText('default slot 2');
+
+    const results7SlotStartChildren = $$('.results7 div article span content-start');
+    await expect(results7SlotStartChildren[0]).toHaveText('start slot 1');
+    await expect(results7SlotStartChildren[1]).toHaveText('start slot 2');
+
+    const results7DefaultSlotChildren = $$('.results7 div content-default');
+    await expect(results7DefaultSlotChildren[0]).toHaveText('default slot 1');
+    await expect(results7DefaultSlotChildren[1]).toHaveText('default slot 2');
+
+    const results8SlotEndChildren = $$('.results8 div section content-end');
+    await expect(results8SlotEndChildren[0]).toHaveText('end slot 1');
+    await expect(results8SlotEndChildren[1]).toHaveText('end slot 2');
+
+    const results9SlotEndChildren = $$('.results9 div section content-end');
+    await expect(results9SlotEndChildren[0]).toHaveText('end slot 1');
+    await expect(results9SlotEndChildren[1]).toHaveText('end slot 2');
+
+    const results9DefaultSlotChildren = $$('.results9 div content-default');
+    await expect(results9DefaultSlotChildren[0]).toHaveText('default slot 1');
+    await expect(results9DefaultSlotChildren[1]).toHaveText('default slot 2');
+
+    const results10Children = $$('.results10 div > *');
+    await expect(results10Children[3]).toHaveText('default slot 1');
+    await expect(results10Children[4]).toHaveText('default slot 2');
+    await expect(results10Children[5]).toHaveText('default slot text node');
+
+    const results11 = document.querySelector('.results11 div');
+    expect(results11.childNodes[1].childNodes[0].childNodes[1].textContent.trim()).toBe('start slot 1');
+    expect(results11.childNodes[1].childNodes[0].childNodes[2].textContent.trim()).toBe('start slot 2');
+    expect(results11.childNodes[3].textContent.trim()).toBe('default slot 1');
+    expect(results11.childNodes[4].textContent.trim()).toBe('default slot 2');
+    expect(results11.childNodes[5].textContent.trim()).toBe('default slot text node');
+    expect(results11.childNodes[6].childNodes[1].textContent.trim()).toBe('end slot 1');
+    expect(results11.childNodes[6].childNodes[2].textContent.trim()).toBe('end slot 2');
+
+    const results12 = document.querySelector('.results12 div');
+    expect(results12.childNodes[1].childNodes[0].childNodes[1].textContent.trim()).toBe('start slot 1');
+    expect(results12.childNodes[1].childNodes[0].childNodes[2].textContent.trim()).toBe('start slot 2');
+    expect(results12.childNodes[3].textContent.trim()).toBe('default slot text node');
+    expect(results12.childNodes[4].textContent.trim()).toBe('default slot 1');
+    expect(results12.childNodes[5].textContent.trim()).toBe('default slot 2');
+    expect(results12.childNodes[6].childNodes[1].textContent.trim()).toBe('end slot 1');
+    expect(results12.childNodes[6].childNodes[2].textContent.trim()).toBe('end slot 2');
```

Comment:
How about this (imho more simplified) approach of validating what text content is visible. It removes the need of jumping through child nodes and makes the test more readable. WebDriver's `getText` command ensures that only visible is being returned which is I think exactly what we are trying to validate here: checking the order and visibility of child nodes.

```suggestion
    await expect($('.results2 div')).toHaveText(
      'default slot text node'
    )

    await expect($('.results3 div')).toHaveText(
      'default slot element 1default slot element 2default slot element 3'
    )

    await expect($('.results4 div')).toHaveText(
      'start slot 1start slot 2\n' +
      'default slot'
    )

    await expect($('.results5 div')).toHaveText(
      'start slot 1start slot 2\n' +
      'default text node'
    )

    await expect($('.results6 div')).toHaveText(
      'start slot 1start slot 2\n' +
      'default slot 1default slot 2'
    )

    await expect($('.results7 div')).toHaveText(
      'start slot 1start slot 2\n' +
      'default slot 1default slot 2'
    )

    await expect($('.results8 div')).toHaveText(
      'end slot 1end slot 2'
    )

    await expect($('.results9 div')).toHaveText(
      'default slot 1default slot 2\n' +
      'end slot 1end slot 2'
    )

    await expect($('.results10 div')).toHaveText(
      'default slot 1default slot 2default slot text node\n' +
      'end slot 1end slot 2'
    )

    await expect($('.results11 div')).toHaveText(
      'start slot 1start slot 2\n' +
      'default slot 1default slot 2default slot text node\n' +
      'end slot 1end slot 2'
    )

    await expect($('.results12 div')).toHaveText(
      'start slot 1start slot 2\n' +
      'default slot text nodedefault slot 1default slot 2\n' +
      'end slot 1end slot 2'
    )
```

Reply:
> How about this (imho more simplified) approach of validating what text content is visible. It removes the need of jumping through child nodes and makes the test more readable. WebDriver's getText command ensures that only visible is being returned which is I think exactly what we are trying to validate here: checking the order and visibility of child nodes.

I think the hierarchy is also important here too though - since we're emulating slots in a non-shadow dom environment, if I understand this test correctly, it's also important that we assert that the text nodes are children of the correct parents here